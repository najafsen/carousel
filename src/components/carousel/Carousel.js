import React, { memo, useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import "./Carousel.scss";

export const Carousel = memo(
  ({
    children,
    interval = 4000,
    transition = 500,
    timingFunction = "ease-in-out"
  }) => {
    const rootRef = useRef(null);
    const timeout = useRef(null);
    const [currentIndex, setCurrentIndex] = useState(0);

    /**
     * When one of the given dependencies change, update marginLeft, clear previous timeout and create new timeout
     */
    useEffect(() => {
      rootRef.current.children[0].style.marginLeft = `-${currentIndex}00%`;

      clearTimeout(timeout.current);
      timeout.current = setTimeout(() => {
        setCurrentIndex((currentIndex + 1) % children.length);
      }, interval);

      return () => clearTimeout(timeout.current);
    }, [currentIndex, children.length, interval]);

    /**
     * when one of dependencies change, update transition and timing function of slides
     */
    useEffect(() => {
      Array.from(rootRef.current.children || []).forEach((child, index) => {
        if (index >= children.length) {
          // we are on navigation buttons container
          return;
        }

        child.style.transition = `margin-left ${transition}ms ${timingFunction}`;
      });
    }, [transition, timingFunction, children.length]);

    return (
      <div data-testid="carousel" ref={rootRef} className="carousel-component">
        {children}
        <div data-testid="navigation-buttons" className="navigation-buttons">
          {React.Children.map(children, (child, childIndex) => (
            <div
              className={cx("navigation-button", {
                active: childIndex === currentIndex
              })}
              onClick={() => {
                setCurrentIndex(childIndex);
              }}
            ></div>
          ))}
        </div>
      </div>
    );
  }
);

Carousel.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  interval: PropTypes.number,
  transition: PropTypes.number,
  timingFunction: PropTypes.string
};
