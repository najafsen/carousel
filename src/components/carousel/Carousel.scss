$navigation-button-color: #6e7986;
$navigation-button-color-active: #ffffff;

.carousel-component {
  position: relative;
  display: flex;
  flex: 1;
  overflow: hidden;

  .carousel-slide-component {
    display: flex;
    flex: 1 0 100%;
  }

  .navigation-buttons {
    position: absolute;
    display: flex;
    bottom: 10%;
    left: 137px;
    right: 137px;
    height: 4px;

    .navigation-button {
      flex: 1 1 100%;
      background-color: $navigation-button-color;
      margin: 0 2px;
      cursor: pointer;

      &.active {
        background-color: $navigation-button-color-active;
      }
    }
  }
}
