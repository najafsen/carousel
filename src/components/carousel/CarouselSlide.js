import React, { memo } from "react";
import PropTypes from "prop-types";

export const CarouselSlide = memo(({ children }) => (
  <div className="carousel-slide-component">{children}</div>
));

CarouselSlide.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};
