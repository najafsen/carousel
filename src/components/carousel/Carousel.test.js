import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Carousel } from "./Carousel";
import { CarouselSlide } from "./CarouselSlide";
import { act } from "react-dom/test-utils";

describe("Carousel", () => {
  test("matches snapshot", () => {
    const { asFragment } = render(
      <Carousel>
        <CarouselSlide>slide1</CarouselSlide>
        <CarouselSlide>slide2</CarouselSlide>
        <CarouselSlide>slide3</CarouselSlide>
      </Carousel>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("shows third slide after 10 seconds", () => {
    jest.useFakeTimers();

    const { queryByTestId } = render(
      <Carousel>
        <CarouselSlide>slide1</CarouselSlide>
        <CarouselSlide>slide2</CarouselSlide>
        <CarouselSlide>slide3</CarouselSlide>
        <CarouselSlide>slide4</CarouselSlide>
        <CarouselSlide>slide5</CarouselSlide>
      </Carousel>
    );
    expect(queryByTestId("carousel").children[0].style.marginLeft).toBe(
      "-000%"
    );
    act(() => {
      jest.advanceTimersByTime(4000);
    });
    act(() => {
      jest.advanceTimersByTime(4000);
    });
    act(() => {
      jest.advanceTimersByTime(2000);
    });
    expect(queryByTestId("carousel").children[0].style.marginLeft).toBe(
      "-200%" // margin left -200% means that the third slide is visible
    );
  });

  test("shows forth slide after 10 seconds when interval prop is set to 3000", () => {
    jest.useFakeTimers();

    const { queryByTestId } = render(
      <Carousel interval={3000}>
        <CarouselSlide>slide1</CarouselSlide>
        <CarouselSlide>slide2</CarouselSlide>
        <CarouselSlide>slide3</CarouselSlide>
        <CarouselSlide>slide4</CarouselSlide>
        <CarouselSlide>slide5</CarouselSlide>
      </Carousel>
    );
    expect(queryByTestId("carousel").children[0].style.marginLeft).toBe(
      "-000%"
    );
    act(() => {
      jest.advanceTimersByTime(3000);
    });
    act(() => {
      jest.advanceTimersByTime(3000);
    });
    act(() => {
      jest.advanceTimersByTime(3000);
    });
    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(queryByTestId("carousel").children[0].style.marginLeft).toBe(
      "-300%" // margin left -300% means that the forth slide is visible
    );
  });

  test("shows forth slide by clicking on the forth navigation button", () => {
    const { queryByTestId } = render(
      <Carousel>
        <CarouselSlide>slide1</CarouselSlide>
        <CarouselSlide>slide2</CarouselSlide>
        <CarouselSlide>slide3</CarouselSlide>
        <CarouselSlide>slide4</CarouselSlide>
        <CarouselSlide>slide5</CarouselSlide>
      </Carousel>
    );
    expect(queryByTestId("carousel").children[0].style.marginLeft).toBe(
      "-000%"
    );
    fireEvent.click(queryByTestId("navigation-buttons").children[3]);
    expect(queryByTestId("carousel").children[0].style.marginLeft).toBe(
      "-300%" // margin left -300% means that the forth slide is visible
    );
  });

  test("have initial transition and timingFunction styles", () => {
    const { queryByTestId } = render(
      <Carousel>
        <CarouselSlide>slide1</CarouselSlide>
        <CarouselSlide>slide2</CarouselSlide>
        <CarouselSlide>slide3</CarouselSlide>
        <CarouselSlide>slide4</CarouselSlide>
        <CarouselSlide>slide5</CarouselSlide>
      </Carousel>
    );
    for (let i = 0; i < 5; i++) {
      expect(queryByTestId("carousel").children[i].style.transition).toBe(
        "margin-left 500ms ease-in-out"
      );
    }
  });

  test("have given transition and timingFunction styles", () => {
    const { queryByTestId } = render(
      <Carousel transition={300} timingFunction="linear">
        <CarouselSlide>slide1</CarouselSlide>
        <CarouselSlide>slide2</CarouselSlide>
        <CarouselSlide>slide3</CarouselSlide>
        <CarouselSlide>slide4</CarouselSlide>
        <CarouselSlide>slide5</CarouselSlide>
      </Carousel>
    );
    for (let i = 0; i < 5; i++) {
      expect(queryByTestId("carousel").children[i].style.transition).toBe(
        "margin-left 300ms linear"
      );
    }
  });

  // More tests are required, but because of lack of time, I tried to write some samples including different scenarios
});
