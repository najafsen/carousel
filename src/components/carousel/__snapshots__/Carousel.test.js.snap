// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="carousel-component"
    data-testid="carousel"
  >
    <div
      class="carousel-slide-component"
      style="margin-left: -000%; transition: margin-left 500ms ease-in-out;"
    >
      slide1
    </div>
    <div
      class="carousel-slide-component"
      style="transition: margin-left 500ms ease-in-out;"
    >
      slide2
    </div>
    <div
      class="carousel-slide-component"
      style="transition: margin-left 500ms ease-in-out;"
    >
      slide3
    </div>
    <div
      class="navigation-buttons"
      data-testid="navigation-buttons"
    >
      <div
        class="navigation-button active"
      />
      <div
        class="navigation-button"
      />
      <div
        class="navigation-button"
      />
    </div>
  </div>
</DocumentFragment>
`;
